File: C# .Net/JDI UI Framework/JDI/Web/Selenium/Elements/Complex/table/Cell.cs
using System;
using Epam.JDI.Commons;
using Epam.JDI.Core;
using Epam.JDI.Web.Selenium.DriverFactory;
using Epam.JDI.Web.Selenium.Elements.Base;
using Epam.JDI.Web.Selenium.Elements.Complex.table.interfaces;
using Epam.JDI.Web.Settings;
using OpenQA.Selenium;

namespace Epam.JDI.Web.Selenium.Elements.Complex.table
{
    class Cell : OptionElement, ICell
    {
        public string Value { get; set; }
        public int ColumnIndex { get; }
        public int RowIndex { get; }
        public int ColumnNum { get; set; }
        public int RowNum
        {
            get
            {
                throw new NotImplementedException();
            }
            set { _rowNum = value; }
        }
        public string ColumnName { get; set; }
        public string RowName { get; set; }
        public By CellLocatorTemplate { get; } = By.XPath(".//tr[{1}]/td[{0}]"); // TODO
        public Table Table { get; }
        public new IWebElement WebElement { get; set; } // TODO
        private int _rowNum;

        public Cell(IWebElement webElement, int columnNum, int rowNum, string columnName, string rowName, By cellLocatorTemplate, Table table)
        {
            WebElement = webElement;
            ColumnNum = columnNum;
            RowNum = rowNum;
            ColumnName = columnName;
            RowName = rowName;
            if (CellLocatorTemplate != null)
                CellLocatorTemplate = cellLocatorTemplate;
            Table = table;
        }

        public Cell(int columnIndex, int rowIndex, int columnNum, int rowNum, string columnName, string rowName, By cellLocatorTemplate, Table table)
        {
            ColumnIndex = columnIndex;
            RowIndex = rowIndex;
            ColumnNum = columnNum;
            RowNum = rowNum;
            ColumnName = columnName;
            RowName = rowName;
            if (CellLocatorTemplate != null)
                CellLocatorTemplate = cellLocatorTemplate;
            Table = table;
        }

        protected new string GetTextAction()
        {
            return Get().Text;
        }
        protected new void ClickAction()
        {
            Get().Click();
        }
        protected new bool IsSelectedAction() // TODO
        {
            return Get().Selected;
        }

        public OptionElement Get()
        {
            return WebElement != null
                    ? new OptionElement(webElement: WebElement)
                    : new OptionElement(CellLocatorTemplate.FillByTemplate(ColumnIndex, RowIndex));
        }

        public T Get<T>(Type clazz) where T : WebBaseElement
        {
            var instance = ExceptionUtils.ActionWithException(() => (T) (clazz.IsInterface
                ? (T) Activator.CreateInstance(MapInterfaceToElement.ClassFromInterface(clazz))
                : Activator.CreateInstance(clazz)), ex => $"Can't get Cell from interface/class: {clazz.Name} .");
            return Get(instance);
        }

        private T Get<T>(T cell) where T : WebBaseElement
        {
            By locator = null; //cell.Locator; TODO Locator missong
            if (locator == null || locator.ToString().Equals(""))
                locator = CellLocatorTemplate;
            if (!locator.ToString().Contains("{0}") || !locator.ToString().Contains("{1}"))
                throw new Exception("Can't create cell with locator template " /*+ cell.Locator */+ // TODO Locator missong
                        ". Template for Cell should contains '{0}' - for column and '{1}' - for row indexes.");
            cell.WebAvatar.ByLocator = WebDriverByUtils.FillByTemplate(locator, RowIndex, ColumnIndex);
            //cell.Avatar.Context.Add(ContextType.Locator, table.Locator); // TODO table.Locator missing
            return cell;
        }

        public ICell UpdateData(string colName, string rowName)
        {
            if ((ColumnName == null || ColumnName.Equals("")) && !(colName == null || colName.Equals("")))
                ColumnName = colName;
            if ((RowName == null || RowName.Equals("")) && !(rowName == null || rowName.Equals("")))
                RowName = rowName;
            return this;
        }
    }
}

File: jdi-uitest-win/src/main/java/com/epam/jdi/uitests/win/winnium/elements/complex/table/Cell.java
package com.epam.jdi.uitests.win.winnium.elements.complex.table;

import com.epam.commons.LinqUtils;
import com.epam.jdi.uitests.core.interfaces.MapInterfaceToElement;
import com.epam.jdi.uitests.core.interfaces.base.IBaseElement;
import com.epam.jdi.uitests.core.interfaces.base.ISelect;
import com.epam.jdi.uitests.core.interfaces.complex.interfaces.ICell;
import com.epam.jdi.uitests.core.settings.JDISettings;
import com.epam.jdi.uitests.win.winnium.elements.WinCascadeInit;
import com.epam.jdi.uitests.win.winnium.elements.apiInteract.GetElementModule;
import com.epam.jdi.uitests.win.winnium.elements.base.Element;
import com.epam.jdi.uitests.win.winnium.elements.base.SelectElement;
import com.epam.jdi.uitests.win.winnium.elements.base.managers.ClickableManager;
import com.epam.jdi.uitests.win.winnium.elements.base.managers.WebElementTextManager;
import org.openqa.selenium.By;

public class Cell extends Element implements ICell {
    private int columnNum, rowNum;
    private String columnName, rowName;
    private Table table;
    private WebElementTextManager textManager = new WebElementTextManager(this);
    private ClickableManager clickableManager = new ClickableManager(this);

    Cell(int columnIndex, int rowIndex, int columnNum, int rowNum, String colName, String rowName,
         String xpathTemplate, Table table) {
        this.columnNum = columnNum;
        this.rowNum = rowNum;
        this.columnName = colName;
        this.rowName = rowName;
        this.table = table;

        this.setParent(table);

        String xpathExpression = String.format(xpathTemplate, rowIndex, columnIndex);
        this.setAvatar(new GetElementModule(By.xpath(xpathExpression), this));
    }

    @Override
    public ISelect get() {
        SelectElement selectElement = new SelectElement(getWebElement());
        selectElement.setParent(table);
        return selectElement;
    }

    @Override
    public <T extends IBaseElement> T get(Class<T> clazz) {
        T instance;
        try {
            instance = (clazz.isInterface())
                    ? (T) MapInterfaceToElement.getClassFromInterface(clazz).newInstance()
                    : clazz.newInstance();
        } catch (Exception ex) {
            throw JDISettings.exception("Can't get Cell from interface/class: " +
                    LinqUtils.last((clazz + "").split("\\.")));
        }
        return get(instance);
    }

    @Override
    public <T extends IBaseElement> T get(T element) {
        element.setParent(this);

        new WinCascadeInit().initElements(element, JDISettings.driverFactory.currentDriverName());

        return element;
    }

    @Override
    public int columnNum() {
        return columnNum;
    }

    @Override
    public int rowNum() {
        return rowNum;
    }

    @Override
    public String columnName() {
        return (columnName != null && !columnName.equals(""))
                ? columnName
                : table.columns().headers().get(columnNum - 1);
    }

    @Override
    public String rowName() {
        return (rowName != null && !rowName.equals(""))
                ? rowName
                : table.rows().headers().get(rowNum - 1);
    }

    @Override
    public void select() {
        click();
    }

    @Override
    public boolean isSelected() {
        return get().isSelected();
    }

    @Override
    public void click() {
        clickableManager.click();
    }

    @Override
    public String getText() {
        return textManager.getText();
    }

    @Override
    public String waitText(String text) {
        return textManager.waitText(text);
    }

    @Override
    public String waitMatchText(String regEx) {
        return textManager.waitMatchText(regEx);
    }

    @Override
    public String getValue() {
        return getText();
    }
}
